library/src/main/scala-new-collections/scala/scalajs/js/WrappedDictionary.scala

```diff
@@ -0,0 +1,162 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js API               **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013, LAMP/EPFL        **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-lang.org/     **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+package scala.scalajs.js
+
+import scala.collection.mutable
+import scala.collection.mutable.Builder
+
+import scala.scalajs.js
+import scala.scalajs.js.annotation._
+
+/** Wrapper to use a js.Dictionary as a scala.mutable.Map */
+@inline
+class WrappedDictionary[A](val dict: js.Dictionary[A])
+    extends mutable.AbstractMap[String, A]
+    with mutable.MapOps[String, A, mutable.Map, WrappedDictionary[A]] {
+
+  import WrappedDictionary._
+
+  protected[this] override def fromSpecificIterable(
+      coll: scala.collection.Iterable[(String, A)]
+  ): WrappedDictionary[A] = {
+    val d = WrappedDictionary.empty[A]
+    d ++= coll
+    d
+  }
+
+  protected[this] override def newSpecificBuilder: Builder[(String, A), WrappedDictionary[A]] =
+    new WrappedDictionary.WrappedDictionaryBuilder[A]
+
+  def get(key: String): Option[A] = {
+    if (contains(key))
+      Some(rawApply(key))
+    else
+      None
+  }
+
+  override def apply(key: String): A = {
+    if (contains(key))
+      rawApply(key)
+    else
+      throw new NoSuchElementException("key not found: " + key)
+  }
+
+  @inline
+  private def rawApply(key: String): A =
+    dict.asInstanceOf[DictionaryRawApply[A]].rawApply(key)
+
+  override def contains(key: String): Boolean = {
+    /* We have to use a safe version of hasOwnProperty, because
+     * "hasOwnProperty" could be a key of this dictionary.
+     */
+    safeHasOwnProperty(dict, key)
+  }
+
+  def subtractOne(key: String): this.type = {
+    if (contains(key))
+      js.special.delete(dict, key)
+    this
+  }
+
+  override def update(key: String, value: A): Unit =
+    dict.update(key, value)
+
+  def addOne(kv: (String, A)): this.type = {
+    dict(kv._1) = kv._2
+    this
+  }
+
+  def clear(): Unit =
+    keysIterator.foreach(subtractOne)
+
+  def iterator: scala.collection.Iterator[(String, A)] =
+    new DictionaryIterator(dict)
+
+  @inline
+  override def keys: scala.collection.Iterable[String] =
+    js.Object.keys(dict.asInstanceOf[js.Object])
+
+  override def empty: WrappedDictionary[A] =
+    new WrappedDictionary(js.Dictionary.empty)
+
+  def map[B](f: ((String, A)) => (String, B)): WrappedDictionary[B] = {
+    val b = new WrappedDictionary.WrappedDictionaryBuilder[B]
+    val it = this.iterator
+    while (it.hasNext) {
+      b += f(it.next())
+    }
+    b.result()
+  }
+
+  // TODO overload collect, flatMap and concat
```

Comment:
@julienrf How import is this TODO? What are we missing if we do not do it now? And how hard would it be to address?

Reply:
Oh, sorry I forgot about that TODO. If we don’t overload `collect`, `flatMap` and `concat` they can only return a `mutable.Map` (since we don’t rely anymore on `CanBuildFrom` to determine the result type. Above you can see that I’ve added an overload of `map` that returns a `WrappedDictionary[B]` when the transformation function returns a tuple `(String, B)`.

Reply:
OK, thanks. We should probably address that, then, because it might be a source breaking change for some clients.

Reply:
Done.